homeassistant/components/device_tracker/ddwrt.py

```diff
@@ -0,0 +1,162 @@
+                        mac_index = (idx * 5) + 2
+                        if mac_index < len(elements):
+                            mac = elements[mac_index]
+                            self.mac2name[mac] = elements[idx * 5]
+
+            return self.mac2name.get(device, None)
+
+    @Throttle(MIN_TIME_BETWEEN_SCANS)
+    def _update_info(self):
+        """ Ensures the information from the DdWrt router is up to date.
+            Returns boolean if scanning successful. """
+        if not self.success_init:
+            return False
+
+        with self.lock:
+            _LOGGER.info("Checking ARP")
+
+            url = 'http://{}/Status_Wireless.live.asp'.format(self.host)
+            data = self.get_ddwrt_data(url)
+
+            if not data:
+                return False
+
+            if data:
+                self.last_results = []
+                active_clients = data.get('active_wireless', None)
+                if active_clients:
+                    # This is really lame, instead of using JSON the ddwrt UI
+                    # uses it's own data format for some reason and then
+                    # regex's out values so I guess I have to do the same,
+                    # LAME!!!
+
+                    # remove leading and trailing single quotes
+                    clean_str = active_clients.strip().strip("'")
+                    elements = clean_str.split("','")
+
+                    num_clients = int(len(elements)/9)
+                    for idx in range(0, num_clients):
+                        # get every 9th element which is the MAC address
+                        index = idx * 9
+                        if index < len(elements):
+                            self.last_results.append(elements[index])
+
+                    return True
+
+            return False
+
+    def get_ddwrt_data(self, url):
+        """ Retrieve data from DD-WRT and return parsed result  """
+        try:
+            response = requests.get(
+                url,
+                auth=(self.username, self.password),
+                timeout=4)
+        except requests.exceptions.Timeout:
+            _LOGGER.exception("Connection to the router timed out")
+            return
+        if response.status_code == 200:
+            return _parse_ddwrt_response(response.text)
+        elif response.status_code == 401:
+            # Authentication error
+            _LOGGER.exception(
+                "Failed to authenticate, "
+                "please check your username and password")
+            return
+        else:
+            _LOGGER.error("Invalid response from ddwrt: %s", response)
+
+
+def _parse_ddwrt_response(data_str):
+    """ Parse the awful DD-WRT data format, why didn't they use JSON????.
+        This code is a python version of how they are parsing in the JS  """
+    data = {}
+    pattern = re.compile(r'\{(\w+)::([^\}]*)\}')
+    for (key, val) in re.findall(pattern, data_str):
+        data[key] = val
+
+    return data
```

Comment:
Tip: Python has besides list comprehension also support for dictionary comprehension. You could rewrite this method as:
```python
return {key: val for key, val in re.compile(r'\{(\w+)::([^\}]*)\}').findall(data_str)}
```